### PR TITLE
Prevent re-use of sessions across retries

### LINKED ISSUE #409: responseClassifier not working with http/1.0 requests
It appears that retries aren't working as expected for http clients that don't use persistent connections when talking to linkerd.

For example, I have the following router config:

```
routers:
- protocol: http
  baseDtab: /http => /$/inet/127.1/9000;
  servers:
  - port: 4140
    ip: 0.0.0.0
  responseClassifier:
    kind: retryableRead5XX
```

Which forwards requests to a go http server running on port 9000.  The go server is configured to respond to 10% of requests with a 503:

```
package main

import (
	"fmt"
	"math/rand"
	"net/http"
)

func main() {
	http.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {
		if rand.Intn(10) == 0 {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		fmt.Fprintf(w, "pong\n")
	})
	http.ListenAndServe(":9000", nil)
}
```

If I send 1000 http/1.1 requests to the router with curl, it uses persistent connections by default and all 1000 requests succeed.

```
$ for i in {1..1000}; do curl -so /dev/null -w '%{http_code}\n' :4140/ping; done | sort | uniq -c
1000 200
```

From the router stats:

```
$ curl -s :9990/admin/metrics.json | jq -S . | grep -E '(success|failures)' | grep -v zipkin
  "rt/http/dst/id/$/inet/127.1/9000/failures": 119,
  "rt/http/dst/id/$/inet/127.1/9000/failures/com.twitter.finagle.service.ResponseClassificationSyntheticException": 119,
  "rt/http/dst/id/$/inet/127.1/9000/success": 1000,
  "rt/http/srv/0.0.0.0/4140/success": 1000,
```

You can see that roughly 10% of client requests failed, but overall router success rate was 100%.  This is confirmed by looking at requeue stats:

```
$ curl -s :9990/admin/metrics.json | jq -S . | grep requeues | grep -v zipkin
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues": 119,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.avg": 0.119,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.count": 1000,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.max": 3,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.min": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p50": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p90": 1,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p95": 1,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p99": 1,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p9990": 3,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p9999": 3,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.sum": 119,
```

If instead I send http/1.0 requests that don't use persistent connections, the retry behavior does not work as expected, and I see 502 and 503 responses from the router.

```
$ for i in {1..1000}; do curl -0 -so /dev/null -w '%{http_code}\n' :4140/ping; done | sort | uniq -c
 890 200
   3 502
 107 503
```

The router stats show an overall router success rate of 89%:

```
$ curl -s :9990/admin/metrics.json | jq -S . | grep -E '(success|failures)' | grep -v zipkin
  "rt/http/dst/id/$/inet/127.1/9000/failures": 122,
  "rt/http/dst/id/$/inet/127.1/9000/failures/com.twitter.finagle.ChannelClosedException": 12,
  "rt/http/dst/id/$/inet/127.1/9000/failures/com.twitter.finagle.service.ResponseClassificationSyntheticException": 110,
  "rt/http/dst/id/$/inet/127.1/9000/sourcedfailures/$/inet/127.1/9000": 9,
  "rt/http/dst/id/$/inet/127.1/9000/sourcedfailures/$/inet/127.1/9000/com.twitter.finagle.ChannelClosedException": 9,
  "rt/http/dst/id/$/inet/127.1/9000/success": 890,
  "rt/http/srv/0.0.0.0/4140/failures": 110,
  "rt/http/srv/0.0.0.0/4140/failures/com.twitter.finagle.ChannelClosedException": 3,
  "rt/http/srv/0.0.0.0/4140/failures/com.twitter.finagle.service.ResponseClassificationSyntheticException": 107,
  "rt/http/srv/0.0.0.0/4140/sourcedfailures/$/inet/127.1/9000": 3,
  "rt/http/srv/0.0.0.0/4140/sourcedfailures/$/inet/127.1/9000/com.twitter.finagle.ChannelClosedException": 3,
  "rt/http/srv/0.0.0.0/4140/success": 890,
```

And the requeue stats indicate that some requests were retried 4 times and still failed, which is unlikely:

```
$ curl -s :9990/admin/metrics.json | jq -S . | grep requeues | grep -v zipkin
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues": 12,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.avg": 0.012,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.count": 1000,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.max": 4,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.min": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p50": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p90": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p95": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p99": 0,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p9990": 4,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.p9999": 4,
  "rt/http/dst/id/$/inet/127.1/9000/retries/requeues_per_request.sum": 12,
```

The 3 502s that we received correspond to 3 ChannelClosedExceptions in the router's logs.  For instance:

```
I 0517 00:39:50.516 THREAD67: FailureAccrualFactory marking connection to "$/inet/127.1/9000" as dead. Remote Address: Inet(/127.0.0.1:9000,Map())
E 0517 00:39:50.521 THREAD67: service failure
com.twitter.finagle.ChannelClosedException: ChannelException at remote address: /127.0.0.1:9000 from service: 0.0.0.0/4140. Remote Info: Upstream Address: /127.0.0.1:52722, Upstream Client Id: Not Available, Downstream Address: /127.0.0.1:9000, Downstream Client Id: $/inet/127.1/9000, Trace Id: 57efcb7401aafc5e.32b071322c81ef00<:57efcb7401aafc5e
	at com.twitter.finagle.NoStacktrace(Unknown Source)
```

As far as I can tell, when a request to linkerd is made without a persistent http connection and the first response it receives from the downstream is a failure, it goes into a retry loop that fully exhausts the retry budget, and still fails the request.

### PR DESCRIPTION
**Pull request by olix0r**
The retry filter may issue multiple requests on the underlying service factory.
This interaction needs to be managed by a FactoryToService module so that the
underlying service (connection) is released to finagle between requests so
that, for instance, it may be closed in the case of HTTP/1.0 requests.

Fixes #409

### CODE REVIEW

File: router/core/src/main/scala/io/buoyant/router/Router.scala [hunk 1]
```diff
@@ -284,7 +284,22 @@ object StackRouter {
   }
 
   def newPathStack[Req, Rsp]: Stack[ServiceFactory[Req, Rsp]] = {
+    /*
+     * The ordering here is very important:
+     *
+     * - At the top of the path stack, we measure tracing and stats so
+     *   that we have a logical view of the request.  Success rate as
+     *   computed from the stats.
```
Comment (klingerf) [inline router/core/src/main/scala/io/buoyant/router/Router.scala]: nit: incomplete last sentence here

**Comment by klingerf** [comment]
⭐  Can confirm that this fixes the original issue I reported, and look at those gorgeous stats:

```
$ curl -s :9990/admin/metrics.json | jq -S . | grep -E '(success|failures)' | grep -v zipkin
  "rt/http/dst/id/$/inet/127.1/9000/failures": 234,
  "rt/http/dst/id/$/inet/127.1/9000/failures/com.twitter.finagle.service.ResponseClassificationSyntheticException": 234,
  "rt/http/dst/id/$/inet/127.1/9000/success": 2000,
  "rt/http/dst/path/http/1.0/GET/success": 1000,
  "rt/http/dst/path/http/1.1/GET/:4140/success": 1000,
  "rt/http/srv/0.0.0.0/4140/success": 2000,
```

File: router/core/src/main/scala/io/buoyant/router/Router.scala [hunk 2]
```diff
@@ -284,7 +285,29 @@ object StackRouter {
...
+     * - Application-level retries are controlled by [[ClassifiedRetries]].
+     *
+     * - Then, factoryToService is used to manage properly manage
+     *   sessions. We need to ensure that the underlying factory
+     *   (provided by the lower stacks) is provisioned for each
+     *   request to accomdate terminated requests (e.g. HTTP/1.0)
+     *
+     * - The failureRecording module records errors encountered when
+     *   acquiring a service from the underlying service factory. This
+     *   must be installed below factory to service in order to catch
```
Comment (adleong) [inline router/core/src/main/scala/io/buoyant/router/Router.scala]: nit: factoryToService

**Comment by adleong** [comment]
:star: 🏭

**Comment by adleong** [comment]
Looks good.  Would feel more comfortable if there was a test for the dst binding factory change.

File: router/core/src/test/scala/com/twitter/finagle/buoyant/DstBindingFactoryTest.scala [hunk 3]
```diff
@@ -119,20 +130,29 @@ class DstBindingFactoryTest extends FunSuite with Awaits with Exceptions {
...
-    val path1exception = intercept[NoBrokersAvailableException] {
-      await(cache(Dst.Path(Path.read("/usa/ca/la"), Dtab.empty, Dtab.empty)))
-    }
+    val baseDtab = Dtab.read("/usa => /people")
+    val localDtab = Dtab.read("/usa => /money")
 
-    val path2exception = intercept[NoBrokersAvailableException] {
-      await(cache(Dst.Path(Path.read("/usa/ca/sf"), Dtab.empty, Dtab.empty)))
+    val e0 = intercept[NoBrokersAvailableException] {
+      val svc = await(cache(Dst.Path(Path.read("/usa/ca/la"), baseDtab, localDtab)))
```
Comment (adleong) [inline router/core/src/test/scala/com/twitter/finagle/buoyant/DstBindingFactoryTest.scala]: TIOLI: could flatMap instead of awaiting twice

**Comment by adleong** [comment]
:star: LGTM

**Pull request merged**